windows-agent/internal/daemon/daemon_test.go

```diff
@@ -144,35 +145,73 @@ func TestStartQuit(t *testing.T) {
 	}
 }
 
-func TestServeNoWSLIP(t *testing.T) {
-	// This test is not parallel because it modifies a global flag that
-	// affects the behavior of the getWslIP function.
-	daemon.SetWslIPErr(t)
-
-	addrDir := t.TempDir()
+func TestServeWSLIP(t *testing.T) {
+	t.Parallel()
 
 	registerer := func(context.Context) *grpc.Server {
 		return grpc.NewServer()
 	}
 
-	// Very lenient timeout because we expect the Serve to fail immediately.
-	// If it doesn't, the test will fail due to the context timeout (otherwise
-	// it would hang indefinitely).
-	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
-	defer cancel()
+	testcases := map[string]struct {
+		netmode             string
+		denyWslSystemAccess bool
+		withAdapters        daemon.MockIPAdaptersState
 
-	d := daemon.New(ctx, registerer, addrDir)
-	defer d.Quit(ctx, false)
+		wantErr bool
+	}{
+		"Success":                       {withAdapters: daemon.Ok},
+		"With a single Hyper-V Adapter": {withAdapters: daemon.SingleHyperVAdapterInList},
+		"With mirrored networking mode": {netmode: "mirrored", withAdapters: daemon.Ok},
+		"With no access to wsl --system but net mode is the default (NAT)": {denyWslSystemAccess: true, withAdapters: daemon.Ok},
+
+		"Error when the networking mode is unknown":       {netmode: "unknown", wantErr: true},
+		"Error when the list of adapters is empty":        {withAdapters: daemon.EmptyList, wantErr: true},
+		"Error when there is no Hyper-V adapter the list": {withAdapters: daemon.NoHyperVAdapterInList, wantErr: true},
+	}
 
-	err := d.Serve(ctx)
-	require.Error(t, err, "Serve should fail when the WSL IP cannot be found")
-
-	select {
-	case <-ctx.Done():
-		// Most likely, Serve did not fail and instead started serving,
-		// only to be stopped by the test timeout.
-		require.Fail(t, "Serve should have failed immediately")
-	default:
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			addrDir := t.TempDir()
+			// Very lenient timeout because we either expect Serve to fail immediately or we stop it manually.
+			// As the last resource, the test will fail due to the context timeout (otherwise it would hang indefinitely).
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+			defer cancel()
+
+			d := daemon.New(ctx, registerer, addrDir)
+
+			wsl := daemon.NewWslSystemMock(tc.netmode, nil, tc.denyWslSystemAccess)
+			ipconfig := daemon.NewHostIPConfigMock(tc.withAdapters)
+			serveErr := make(chan error)
+			go func() {
+				serveErr <- d.Serve(ctx, daemon.WithWslSystemDistro(wsl), daemon.WithHostIPConfig(&ipconfig))
+			}()
+
+			if tc.wantErr {
+				require.NotErrorIs(t, <-serveErr, grpc.ErrServerStopped, "Serve should fail when the WSL IP cannot be found")
```

Comment:
shouldn’t we quit the daemon if we fail here? If we can double quit (to be checked) without any issue, I would still `defer d.Quit(ctx, false)` after creating the daemon.

Reply:
Added a `defer d.Quit` right after creating the deamon.